Add Dockerfile

hasufell opened the pull request:
See https://github.com/input-output-hk/cardano-transactions/issues/22 for discussion and decisions

Commit:
Add installation instructions to README

Commit:
Add Dockerfile wrt #22

KtorZ commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+  curl -L -O https://github.com/input-output-hk/cardano-transactions/releases/download/$VERSION/cardano-transactions-$VERSION-linux64.tar.gz && \
+  tar -C /usr -xvzf cardano-transactions-$VERSION-linux64.tar.gz bin/cardano-tx && \
+  chmod +x /usr/bin/cardano-tx && \
+  rm cardano-transactions-$VERSION-linux64.tar.gz
```
That won't do, the goal of the Dockerfile would be to build things locally, not to download a pre-existing binary. 
Pretty much in the lights of: https://github.com/input-output-hk/cardano-addresses/blob/master/Dockerfile

Then, cabal, stack, chose your weapon. However I'd still favor stack here for consistency with cardano-addresses.

hasufell replied:
Alright, that's possible too. I remember that Docker wizards don't really like that approach, because it's very hard to debug (similar to CI) on failures, but I guess we won't really expect many failures due to stack etc.

KtorZ replied:
Yes. Most of our users who build with Docker do it because they want to build from sources without caring about the system dependencies. But the ability to build from sources is essential. 

Arguably, Nix solves this in a nicer way, but Docker is a good answer that we ought to support too.

hasufell replied:
Then I think we should consider caching dependencies on a public S3 bucket: https://github.com/fpco/cache-s3

KtorZ replied:
Possibly as a second iteration. Initially, just leveraging the local docker cache would be enough.

hasufell replied:
Stack here takes about 30~ minutes to clone all source repos (and clones one repo multiple times). Cabal needs about 2 minutes max.

hasufell replied:
After 1 hour of cloning and bulding, I get:

```
--  While building package cardano-crypto-class-2.0.0 using:
      /root/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.5 --builddir=.stack-work/dist/x86_64-linux/Cabal-2.4.0.1 configure --user --package-db=clear --package-db=global --package-db=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/pkgdb --libdir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/lib --bindir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/bin --datadir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/share --libexecdir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/libexec --sysconfdir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/etc --docdir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/doc/cardano-crypto-class-2.0.0 --htmldir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/doc/cardano-crypto-class-2.0.0 --haddockdir=/root/.stack/snapshots/x86_64-linux/8ed63663e0a851b03dbe84e9d9e3c3fd980f39b6217518155664a844f84239d4/8.6.5/doc/cardano-crypto-class-2.0.0 --dependency=aeson=aeson-1.4.6.0-EBlgS3B9Tq36MQnW8dyprT --dependency=base=base-4.12.0.0 --dependency=base16-bytestring=base16-bytestring-0.1.1.6-17atVnUhDnM13pAkKbwL6I --dependency=bytestring=bytestring-0.10.8.2 --dependency=cardano-binary=cardano-binary-1.5.0-5RfPmOHrT8m16CMfVBBr0T --dependency=cardano-prelude=cardano-prelude-0.1.0.0-GlPHUxhp48Z3krZNP7mUog --dependency=cryptonite=cryptonite-0.25-GRKCG37cD79GVqvGPiaNxJ --dependency=deepseq=deepseq-1.4.4.0 --dependency=integer-gmp=integer-gmp-1.0.2.0 --dependency=memory=memory-0.14.18-CTT9ZW74V7DJG216YjnXgb --dependency=text=text-1.2.4.0-GrUF4lk2hk02X2GpHbr2GS --dependency=transformers=transformers-0.5.6.2 --dependency=vector=vector-0.12.1.2-3ubVJHVYUZgKP2BOIpfTrf -f-development --exact-configuration --ghc-option=-fhide-source-paths
    Process exited with code: ExitFailure 1
```

This is why Dockerfile s**x for building.

Commit:
Rather make Dockerfile build from source

Commit:
Add cabal support

Commit:
Use cabal in Dockerfile instead of stack

This decreases build time from ~60 to ~20 minutes,
because cabal doesn't re-clone existing remote git
repos for every subdirectory. Inside docker, networking
is a mess, so this might add to that problem.

hasufell commented:
I have included a commit that adds a github workflow for building and pushing images to dockerhub. For that to actually work we'd need to:

1. create a dedicated "machine" user on our dockerhub inputoutput organisation
2. have devops manage the credentials
3. create an API token on dockerhub for that user
4. add the credentials to `secrets` in this repository settings (using the variable names `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN`)

hasufell commented on README.md:
````diff
@@ -14,6 +14,28 @@
+Run:
+
+```sh
+docker pull inputoutput/cardano-tx
````
maybe not

paweljakubas commented:
I managed to produce image : 
```
pawel@pure:~/Work/IOHK/cardano-transactions (hasufell/ADP-290/create-docker-image)$ docker build --tag  cardano-transactions .
```

paweljakubas commented:
@hasufell maybe I am doing something wrong when running docker image? nevertheless, would be good to add 
 instructions how to built and run an example in README.md

paweljakubas commented:
Now works for me:
```
pawel@pure:~/Work/IOHK/cardano-transactions (hasufell/ADP-290/create-docker-image)$ docker run --rm -ti --entrypoint sh cardano-transactions
/ # cardano-tx empty 764824073 \
> | cardano-tx add-input 0 3b40265111d8bb3c3c608d95b3a0bf83461ace32d79336579a1939b3aad1c0b7 \
> | cardano-tx add-output 42 Ae2tdPwUPEZETXfbQxKMkMJQY1MoHCBS7bkw6TmhLjRvi9LZh1uDnXy319f \
> | cardano-tx lock \
> | cardano-tx sign-with e0860dab46f13e74ab834142e8877b80bf22044cae8ebab7a21ed1b8dc00c155f6b78eee2a5bbd453ce7e7711b2964abb6a36837e475271f18ff36ae5fc8af73e25db39fb78e74d4b53fb51776d0f5eb360e62d09b853f3a87ac25bf83
4ee1fb \
> | cardano-tx serialize
goOfggDYGFgkglggO0AmURHYuzw8YI2Vs6C/g0YazjLXkzZXmhk5s6rRwLcA/5+CgtgYWCGDWByw5pPLyXJyvUKqzztEPsrGi5a0+iAVkYako4iDoAAaSXwVxhgq/6CBggDYGFiFglhAjmICNbKkJyWcwaHGOvnvF5yztuOW2hguAdZd0+176UviXbOft4501LU/tRd20PXrNg5i0JuFPzqHrCW/g07h+1hANoRQGiE0a6PBmm3DR5GA6AoU5e3A/7SxRpq+AzLn4crykdUMuNLEfuN/a0S3+/zFobAbwmIDjpOWm5pyiis1AA==/ # ^C
/ # exit
```

@KtorZ  @rvl  so from my user perpective I confirm it works fine. I was able to build locally and  use it. when it comes to whether other parts (devops) are correct I cannot say and leave it to you

Commit:
Improve docker instructions/documentation

Commit:
Update .dockerignore

Commit:
Use alpine for target image

This is more suitable for interactive shells.

paweljakubas commented:
Now image doubled in size, although still very small:
```
pawel@pure:~/Work/IOHK/cardano-transactions (hasufell/ADP-290/create-docker-image)$ docker images
REPOSITORY                 TAG                 IMAGE ID            CREATED             SIZE
cardano-tx                 latest              572c49fc00ee        3 minutes ago       8.38MB
```
Contener works:
```
pawel@pure:~/Work/IOHK/cardano-transactions (hasufell/ADP-290/create-docker-image)$ docker run --rm -ti --entrypoint sh cardano-tx
/ # cardano-tx empty 764824073 \
> | cardano-tx add-input 0 3b40265111d8bb3c3c608d95b3a0bf83461ace32d79336579a1939b3aad1c0b7 \
> | cardano-tx add-output 42 Ae2tdPwUPEZETXfbQxKMkMJQY1MoHCBS7bkw6TmhLjRvi9LZh1uDnXy319f \
> | cardano-tx lock \
> | cardano-tx sign-with e0860dab46f13e74ab834142e8877b80bf22044cae8ebab7a21ed1b8dc00c155f6b78eee2a5bbd453ce7e7711b2964abb6a36837e475271f18ff36ae5fc8af73e25db39fb78e74d4b53fb51776d0f5eb360e62d09b853f3a87ac25bf83
4ee1fb \
> | cardano-tx serialize
goOfggDYGFgkglggO0AmURHYuzw8YI2Vs6C/g0YazjLXkzZXmhk5s6rRwLcA/5+CgtgYWCGDWByw5pPLyXJyvUKqzztEPsrGi5a0+iAVkYako4iDoAAaSXwVxhgq/6CBggDYGFiFglhAjmICNbKkJyWcwaHGOvnvF5yztuOW2hguAdZd0+176UviXbOft4501LU/tRd20PXrNg5i0JuFPzqHrCW/g07h+1hANoRQGiE0a6PBmm3DR5GA6AoU5e3A/7SxRpq+AzLn4crykdUMuNLEfuN/a0S3+/zFobAbwmIDjpOWm5pyiis1AA==/ # 
/ # exit
```

paweljakubas reviewed (approved):
LGTM (in terms of building/using)
